### ISSUE TITLE
Update wording of IterableStorageMap docs

### ISSUE THREAD

**Issue by xlc** [paritytech/substrate#10358]
> /// Enumerate all elements in the map in lexicographical order of the encoded key. If you
	/// alter the map while doing this, you'll get undefined results.

The undefined results wording is very uncomfortable to me as someone with C++ background. There are no undefined behaviour here. Deleting item is perfectly fine, update value is also ok. Only add item means it may or may not be iterated later depends on the hash value of the key. Can we update the docs to make it more clear?

https://github.com/paritytech/substrate/blob/ee8ac3a6c9c9292911afbf5ec16fbb2bb0d44b27/frame/support/src/storage/mod.rs#L326-L351

**Comment by KiChjang** [comment]
I don't get how iterator invalidation is _not_ undefined behaviour? The bit where you said "deleting/updating a value is also OK", it's actually not; if you had been iterating on a value whilst modifying it, then the iterator abstraction has been broken and it _is_ undefined behaviour, because the iterator does not expect the item's value to be changed by something else while it exists.

Perhaps you meant that in Substrate, storage iterator invalidation is _implementation-defined behaviour_ instead?

**Comment by xlc** [comment]
Yeah in general, it is undefined behaviour, but for this particular iterator, it is not. The iterator offers a mode to remove items after accessing it. We cannot have any undefined behaviour in warm runtime.
